File: src/core/download_validations.py
# Standard library
from urllib.parse import urlsplit

# Constants
from constants import ALLOW_HOSTS, GOOGLE

# Third-Party libraries
from pytubefix import YouTube
from requests import ConnectionError, Timeout, head
from pytubefix.exceptions import (
    AgeRestrictedError,
    LiveStreamError,
    MembersOnly,
    VideoPrivate,
    VideoRegionBlocked,
    VideoUnavailable,
    LoginRequired,
)

# App logging
from app_logging import LoggingManager

# App enums
from app_enums import LogLevels, ExcelTextLoaderKeys

# i18n
from i18n import ExcelTextLoader

logging_manager = LoggingManager()
excel_text_loader = ExcelTextLoader()


class DownloadValidations:
    """
    Validating various aspects of a YouTube video download request
    """

    @staticmethod
    def validate_non_empty_url(url: str):
        """
        Checks if a URL has been entered.

        Args:
            url (str): URL to be checked if it's not an empty string

        Raises:
            ValueError: Error thrown if the given url is an empty string

        Returns:
            bool: True if the given string is not empty
        """

        if not url:
            logging_manager.write_log(LogLevels.WARNING, "No URL entered")

            raise ValueError(
                excel_text_loader.get_text(ExcelTextLoaderKeys.ERROR_EMPTY_URL)
            )

    @staticmethod
    def check_if_youtube_url(url: str):
        """
        Checks if a given URL is from YouTube.

        Args:
            url (str): URL to check if it is from YouTube

        Raises:
            ValueError: Error thrown if the url is not from YouTube

        Returns:
            bool: True if the url is from YouTube
        """

        host = urlsplit(url).hostname

        if host not in ALLOW_HOSTS:
            logging_manager.write_log(
                LogLevels.WARNING, "The URL is not from YouTube"
            )

            raise ValueError(
                excel_text_loader.get_text(
                    ExcelTextLoaderKeys.ERROR_YOUTUBE_URL
                )
            )

    @staticmethod
    def check_internet_connection() -> bool:
        """
        Check internet connection availability.

        Raises:
            ConnectionError: Error thrown if there is no internet connection

        Returns:
            bool: True if there is internet connection
        """

        try:
            head(GOOGLE, timeout=5)

        except (ConnectionError, Timeout) as e:
            logging_manager.write_log(
                LogLevels.ERROR, f"Internet connection issue: {e}"
            )

            raise ConnectionError(
                excel_text_loader.get_text(ExcelTextLoaderKeys.ERROR_CONNECTION)
            )

        return True

    @staticmethod
    def is_youtube_video_available(url: str) -> bool:
        """
        Check if YouTube video is available

        Args:
            url (str): Video URL to check

        Raises:
            ValueError: Error thrown if the video is not available

        Returns:
            bool: True if video is available
        """

        try:
            YouTube(url).check_availability()

            return True

        except (
            AgeRestrictedError,
            LiveStreamError,
            MembersOnly,
            VideoPrivate,
            VideoRegionBlocked,
            VideoUnavailable,
        ) as e:
            logging_manager.write_log(LogLevels.WARNING, str(e))

            error_message = {
                AgeRestrictedError: excel_text_loader.get_text(
                    ExcelTextLoaderKeys.ERROR_AGE_RESTRICTED
                ),
                LiveStreamError: excel_text_loader.get_text(
                    ExcelTextLoaderKeys.ERROR_LIVE_STREAM
                ),
                MembersOnly: excel_text_loader.get_text(
                    ExcelTextLoaderKeys.ERROR_ONLY_MEMBERS
                ),
                VideoPrivate: excel_text_loader.get_text(
                    ExcelTextLoaderKeys.ERROR_PRIVATE_VIDEO
                ),
                VideoRegionBlocked: excel_text_loader.get_text(
                    ExcelTextLoaderKeys.ERROR_BLOCKED_REGION
                ),
                VideoUnavailable: excel_text_loader.get_text(
                    ExcelTextLoaderKeys.ERROR_DEFAULT
                ),
                LoginRequired: excel_text_loader.get_text(
                    ExcelTextLoaderKeys.ERROR_DEFAULT
                ),
            }

            raise ValueError(error_message.get(type(e))) from e

        except Exception as e:
            logging_manager.write_log(LogLevels.CRITICAL, str(e))

            raise Exception(
                excel_text_loader.get_text(ExcelTextLoaderKeys.ERROR_DEFAULT)
            ) from e

File: lib/core/download_validations.dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:http/http.dart' show ClientException, head;
import 'package:logkeeper/logkeeper.dart' show LogKeeper;
import 'package:youtube_explode_dart/youtube_explode_dart.dart'
    show
        VideoRequiresPurchaseException,
        VideoUnavailableException,
        VideoUnplayableException,
        YoutubeExplode;

import '/constants/hosts.dart' show allowHosts, google;
import '/l10n/app_localizations.dart' show AppLocalizations;
import 'core_utils.dart' show getDenoSolver;

class DownloadValidations {
  static final _instance = DownloadValidations._internal();

  factory DownloadValidations() {
    return _instance;
  }

  DownloadValidations._internal();

  static bool nonEmptyURL({
    required BuildContext context,
    required String? url,
  }) {
    if (url == null || url.isEmpty) {
      LogKeeper.warning('No URL entered');

      throw Exception(AppLocalizations.of(context)!.error_empty_url);
    }

    return true;
  }

  static bool checkIfYoutubeUrl({
    required BuildContext context,
    required String url,
  }) {
    final host = Uri.parse(url).host;

    if (!allowHosts.contains(host)) {
      LogKeeper.warning('The $url URL is not from YouTube.');

      return false;
    }

    return true;
  }

  static Future<bool> checkInternetConnection({
    required BuildContext context,
  }) async {
    try {
      await head(Uri.parse(google)).timeout(Duration(seconds: 5));

      return true;
    } on TimeoutException catch (e) {
      if (!context.mounted) return false;

      LogKeeper.error('Internet connection timeout: $e');

      throw Exception(AppLocalizations.of(context)!.error_connection);
    } on ClientException catch (e) {
      if (!context.mounted) return false;

      LogKeeper.error('Internet connection issue: $e');

      throw Exception(AppLocalizations.of(context)!.error_connection);
    }
  }

  static Future<bool> isYoutubeVideoAvailable({
    required BuildContext context,
    required String url,
  }) async {
    final denoSolver = await getDenoSolver();
    final yt = YoutubeExplode(jsSolver: denoSolver);

    try {
      final video = await yt.videos.get(url);

      if (video.isLive) {
        LogKeeper.warning('Video is a live stream: $url');

        if (!context.mounted) return false;

        throw Exception(AppLocalizations.of(context)!.error_live_stream);
      }

      await yt.videos.streamsClient.getManifest(video.id);

      return true;
    } on VideoUnavailableException {
      if (!context.mounted) return false;

      LogKeeper.error('Video unavailable: $url');

      throw Exception(AppLocalizations.of(context)!.error_unavailable_video);
    } on VideoRequiresPurchaseException {
      if (!context.mounted) return false;

      LogKeeper.error('Video requires purchase: $url');

      throw Exception(AppLocalizations.of(context)!.error_only_members);
    } on VideoUnplayableException catch (e) {
      if (!context.mounted) return false;

      LogKeeper.error('Video unplayable: ${e.message}');

      throw Exception(AppLocalizations.of(context)!.error_unavailable_video);
    } on Exception catch (e) {
      if (!context.mounted) return false;

      LogKeeper.error('Error checking video availability: $e');

      throw Exception(AppLocalizations.of(context)!.error_default);
    } finally {
      yt.close();
    }
  }
}
